SBS: Implementation of new StepByStep module.

DanielvanDamme opened the pull request:
The implementation of Step By Step with things roughly changed:

- View for pattern selection with two buttons "Next" and "Continue"
- No combobox visible when selecting "Next"
- Combobox visible when selecting "Continue"

Use cases "Next"
1. Answer yes to adding to opened solution
2. Implement instruction in the opened new file
3. Click the "Check" button
4. Proceed to next step with "Next" button

1. Answer no to adding to opened solution, choose a save location for the new file
3. Implement instruction in the opened new file
4. Click the "Check" button
5. Proceed to next step with "Next" button

Use case "Continue"
1. Choose from combobox which file is to be checked in the opened solution
2. Click the "Check" button
3. Proceed to next step with "Next" button

Commented most functions for documentation as well as inline comments in the methods.

KnapSac commented:
Some things I found before even looking at the code:
- The initial page with the `Next` and `Continue` buttons is quite unclear, what do these buttons mean? Without any additional knowledge, this doesn't provide a good experience to the user.
- When using `Continue`, the dropdown to select which class/file to use performs very poorly. Why don't we just present a file dialog to the user here?
- When using `Next`, I get an `IndexOutOfRange` exception when I press the `Check` button, but I haven't added any code yet.

KnapSac commented on PatternPal/PatternPal.Extension/Commands/NavigateCommand.cs:
```diff
@@ -37,7 +48,142 @@ public class NavigateCommand< T > : ICommand
+
+            // Save location for project provided by user
+            string folderPath = string.Empty;
+            using (FolderBrowserDialog fdb = new FolderBrowserDialog())
```
Why does the user not get this option when choosing to add the file to the solution? It would be better if that behavior is the same.

DanielvanDamme replied:
The CreateNewWorkFile is now called everytime one wants to create a new work file.

KnapSac commented on PatternPal/PatternPal.Extension/ViewModels/StepByStepInstructionsViewModel.cs:
```diff
@@ -17,6 +17,10 @@ public class StepByStepInstructionsViewModel : ViewModel
 
+        public string FilePath { get; set; }
+
+        public bool SBScontinue { get; }
```
```suggestion
        public bool Continue { get; }
```
As you're already in a class specific to Step by Step, you don't need to include that in the property names here.

DanielvanDamme replied:
This was replaced with an enum of modi.

KnapSac commented on PatternPal/PatternPal.Extension/ViewModels/StepByStepListViewModel.cs:
```diff
@@ -29,17 +30,41 @@ public class StepByStepListViewModel : ViewModel
+            Recognizer test = InstructionSetList.FirstOrDefault();
 
-            GetInstructionSetsResponse response = GrpcHelper.StepByStepClient.GetInstructionSets(new GetInstructionSetsRequest());
+            GetInstructionSetResponse response2 =
```
Don't use generic variable names like `response`, `response2`, ect. Instead use more speaking variable names like `instructionSetsResponse` and `instructionSetResponse`.

DanielvanDamme replied:
Variable names of these have changed among other variable names such as temp, test, and such.

KnapSac commented on PatternPal/PatternPal.Extension/Views/StepByStepInstructionsView.xaml.cs:
```diff
@@ -104,67 +107,92 @@ private void InitializeViewModelAndButtons()
+            else
+            {
+                CheckImplementationButton.IsEnabled = 
+                    _viewModel.CurrentInstruction.FileId != "" || !_viewModel.SBScontinue;
```
```suggestion
                    !string.IsNullOrEmpty(_viewModel.CurrentInstruction.FileId) || !_viewModel.SBScontinue;
```

DanielvanDamme replied:
Checking if the check button should be visible is now done by checking whether the filepath(s) that is going to be used for a check call to the service is non empty.

KnapSac commented on PatternPal/PatternPal.Extension/Views/StepByStepInstructionsView.xaml.cs:
```diff
@@ -104,67 +107,92 @@ private void InitializeViewModelAndButtons()
         }
 
+        /// <summary>
+        /// Given the current step checks whether the implementation 
```
This sentence is incomplete.

DanielvanDamme replied:
☑

KnapSac commented on PatternPal/PatternPal.Extension/Views/StepByStepInstructionsView.xaml.cs:
```diff
@@ -104,67 +107,92 @@ private void InitializeViewModelAndButtons()
+                LoadProject();
 
-                ExpanderResults.ResultsView.ItemsSource = viewModels;
+                foreach (Project project in Projects)
```
I'd to the same thing as the `RecognizerService` here, just pass either the path to a `.cs` file or the path to a `.csproj` file to the background service, you can get the files there, instead of keeping this logic in the extension.

DanielvanDamme replied:
The logic regarding obtaining projects in the ide was replaced by having users provide a filepath(s) regardless. These filepaths are always known and sent to the service in a request.

KnapSac commented on PatternPal/PatternPal.Extension/ViewModels/StepByStepListViewModel.cs:
```diff
@@ -29,17 +30,41 @@ public class StepByStepListViewModel : ViewModel
+            GetInstructionSetsResponse response = 
+                GrpcHelper.StepByStepClient.GetInstructionSets(new GetInstructionSetsRequest());
+
+            List<Recognizer> temFilteredRecognizers = response.Recognizers.Where(r => r is Recognizer.Singleton).ToList();
```
Why are you doing this filtering on the frontend? Even if it's temporary, it would still be better to do it in the background service.

DanielvanDamme replied:
Filtering has been moved to the GetInstructionSet in the StepByStepSerivce.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,211 @@
+        GetInstructionSetsResponse response = new();
+
+        // Obtain all the enum values in the proto.
+        bool initialValue = true;
```
This can be improved by using the `RecognizerRunner.SupportedRecognizers.Keys` collection, which was not available in the past. I'd also move this method to a helper class, so that it can be reused by the `RecognizerService`.

DanielvanDamme replied:
I have made use of the supported recognizers in the runner and filter accordingly using a try catch.

KnapSac commented on Protos/step_by_step.proto:
```diff
@@ -19,16 +22,30 @@ service StepByStepService {
+// GetInstructionSet request message.
+message GetInstructionSetRequest {
+    // Name of the PatternPal
+    patternpal.Recognizer selected_combobox = 1;
```
You're mixing concerns here, why should this message know anything about how the instruction set was selected?

DanielvanDamme replied:
I have moved everything to do with getting the instruction set to the StepByStepInstructionsViewmodel. Instead of the StepByStepListViewModel getting the instruction set based on what was selected in the combox it instead it provides the constructor of the StepByStepInstructionsViewModel with the recognizer that was selected and uses that recognizer to return the instruction from the service.

DanielvanDamme replied:
I have additionally removed the static InstructionSetHolder, almost everything in the service now requires a recognizer to retrieve the instruction. It is now "stateless".

KnapSac commented on Protos/step_by_step.proto:
```diff
@@ -19,16 +22,30 @@ service StepByStepService {
     rpc CheckInstruction (CheckInstructionRequest) returns (CheckInstructionResponse);
+
+    // Sets the filepath of every instruction to the newly created .cs file.
+    rpc SetNewFilePath (SetFilePathRequest) returns (SetFilePathResponse);
```
I'd probably make this a property on the `CheckInstruction` request. This saves a roundtrip to the background service.

DanielvanDamme replied:
This has been moved to the logic responsible for making the CheckInstruction.

KnapSac commented on Protos/step_by_step.proto:
```diff
@@ -54,11 +71,8 @@ message GetSelectableClassesResponse {
-    // Id of the instruction to get, 0-based.
-    int32 instruction_id = 2;
+    // Indexer to determine the current instruction
+    uint32 instructionNumber = 1;
```
```suggestion
    uint32 instruction_number = 1;
```

KnapSac commented on Protos/step_by_step.proto:
```diff
@@ -96,9 +104,18 @@ message CheckInstructionRequest {
 message CheckInstructionResponse {
     // Result of the check of the instruction.
-    patternpal.RecognizeResult result = 1;
+    bool result = 1;
```
Why are changing this to be a `bool`? In the case of a correct implementation this is fine, but what if the user made a mistake? It would be better to show the results, so they can see what they did incorrectly.

DanielvanDamme replied:
I'm having trouble using that RecognizeResult because I do not know what information can be retrieved from the ICheckresult. I have found MatchedNode which is useful and the feedback messages.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,211 @@
+        return Task.FromResult(new SetFilePathResponse());
     }
+    
+    public IRecognizer GetRecognizer(Recognizer recognizer)
+    {
+        IDictionary< Recognizer, IRecognizer > supportedRecognizers =
+            new Dictionary< Recognizer, IRecognizer >();
+
+        Type recognizerType = typeof( IRecognizer );
+
+        // Find all types which derive from `IRecognizer`.
+        foreach (Type type in recognizerType.Assembly.GetTypes().Where(ty => ty != recognizerType && recognizerType.IsAssignableFrom(ty)))
+        {
+            IRecognizer instance = (IRecognizer)Activator.CreateInstance(type)!;
+
+            // Get the mapping from `Recognizer` to `IRecognizer` by invoking the property on the
+            // recognizer which specifies it.
+            supportedRecognizers.Add(
+                (Recognizer)type.GetRuntimeProperty(nameof( IRecognizer.RecognizerType ))!.GetValue(instance)!,
+                instance);
+        }
+        return supportedRecognizers[ recognizer ];
+    }
```
```suggestion
```
Why are you copying this method from `RecognizerRunner`?

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,211 @@
+    }
+}
+
+public static class InstructionSetHolder
```
What is the point of this class? When you receive a `CheckInstruction` request, can't you just lookup the `InstructionSet` to use based on the provided info?

DanielvanDamme replied:
See earlier comment about stateless.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,211 @@
+            selectedNodePath,
+            InstructionSetHolder.instructionSet.Steps[request.InstructionId]);
+        IList<ICheckResult> res = runner.Run(pruneAll: true);
+        NodeCheckResult checkRes = (NodeCheckResult)res[0];
```
This crashes if the list is empty.

DanielvanDamme replied:
I now check whether any of the provided files has a correct implementation, if the list was empty (aka no roots provided to the runner) then the implementation was also correct.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,211 @@
         ServerCallContext context)
     {
-        throw new NotImplementedException();
+        SyntaxGraph graph = new();
```
See comment regarding this in the PR.

DanielvanDamme replied:
This was removed and made more intuitive with a OpenFileDialog where the user can provide which files they want recognized and in the "new" option the user always provides a save location for the new file.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
Look at the lint warnings.

DanielvanDamme commented on PatternPal/PatternPal.Extension/Commands/NavigateCommand.cs:
```diff
@@ -37,7 +48,142 @@ public class NavigateCommand< T > : ICommand
+
+            // Start a new project when the user wants to learn to implement the selected pattern
+            // from the combobox.
+            if (viewModel.GetType() == typeof(StepByStepInstructionsViewModel))
```
This was moved to the StepByStepListViewModel, this is before the InstructionsView is presented to the user.

DanielvanDamme commented on PatternPal/PatternPal.Extension/Commands/NavigateCommand.cs:
```diff
@@ -37,7 +48,142 @@ public class NavigateCommand< T > : ICommand
+                        System.IO.Path.Combine(
+                            System.IO.Path.GetDirectoryName(csTemplateProject.FullName),
+                            fileName);
+                    System.IO.File.WriteAllText(filePath, fileContent);
```
File.Create is used everytime a new file is created.

DanielvanDamme commented:
The use cases have changed in this latest commit.

In both cases one has to provide the design pattern they want to implement or continue working on.

Use case user presses "Implement new pattern"
1. Choose yes or no to add or not add the file to the currently opened solution.
2. Choose the location the file is to be saved to.
3. Implement the instruction.
4. Click check.
5. Next button only becomes available after the check is passed.

Use case user presses "Continue work on pattern"
1. Choose the files in which you implemented the design pattern selected.
2. Click check.
3. Next button only becomes available after the check is passed.

KnapSac commented on PatternPal/PatternPal.Extension/ViewModels/StepByStepListViewModel.cs:
```diff
@@ -20,26 +26,152 @@ public class StepByStepListViewModel : ViewModel
+                else
+                {
+                    MessageBox.Show("No files were provided!");
+                    ofd.ShowDialog();
```
```suggestion
```
This is actually incorrect. You don't store the result of this call, which means that you trigger the `ShowDialog` call on line 81 after the user has closed this dialog, possibly 
resulting in us forgetting the files the user has selected.

DanielvanDamme replied:
I have removed this showDialog.

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,154 @@
+        GetInstructionSetsResponse response = new();
+        
+        IDictionary<Recognizer, IRecognizer> sbsDictionary = 
+            RecognizerRunner.SupportedRecognizers;
```
This doesn't copy the dictionary, so the `Remove` call on line 32 will modify it!

KnapSac commented on PatternPal/PatternPal/Services/StepByStepService.cs:
```diff
@@ -1,41 +1,154 @@
+
+        // Go over all the recognizers supported by the runner and remove those that do not implement
+        // the GenerateStepsList function.
+        foreach (KeyValuePair<Recognizer, IRecognizer> entry in sbsDictionary)
+        {
+            try
+            {
+                entry.Value.GenerateStepsList();
+            }
+            catch
+            {
+                sbsDictionary.Remove(entry.Key);
+            }
+        }
+
+        response.Recognizers.Add(sbsDictionary.Keys);
```
```suggestion
        foreach (
        KeyValuePair<Recognizer, IRecognizer> entry in RecognizerRunner.SupportedRecognizers)
        {
            try
            {
                entry.Value.GenerateStepsList();
            }
            catch
            {
                continue;
            }
            
            response.Recognizers.Add(entry.Key);
        }
```

DanielvanDamme replied:
I have done this differently now with the suggestion.

KnapSac reviewed (changes requested):
This looks good!

I've left a few more comments about some issues I found.

One question I still had, is why the file dialog is different between the new and continue options? In the new case I get a weird small dialog, which is very inconvenient. Whereas in the continue case, I get a regular file selector. Can you maybe make it so you always get the larger file selector? I believe it should be possible to also select folders using that dialog.

And it is currently impossible to know whether you have implemented the last step correctly, as there is no next button which can be activated.

DanielvanDamme commented:
I incorporated the feedback plus improved unusual user behaviour in the stepbysteplistviewmodel. 

Also added a textblock that becomes visible when a step is correctly implemented.

!! Problem !!
There is a problem with one of these:
1. The way I wrap the checks in the GenerateStepsList method in the SingletonRecognizer 
2. The way they are provided to the runner in the CheckInstruction method in the StepByStepService
3. They way I interpret the result from the runner.run method.

Because step 2 and 3 (step 1,2 and 3 are the only one implemented in GenerateStepsList in SingletonRecognizer) are always correct even in empty files.

DanielvanDamme commented:
I have fixed this behavior, it now works as intended. I have added in the latest commit added a suggested fix in NodeCheck (from Matteo) in comments that fixes that a FieldCheck succeeds/is correct inside a ClassCheck when there is no field in that class.

I have also made it visible to the user when the implementation is correct. The next button becomes available as before, but now the user is also presented with a "Correct!" when the implementation is correct.

I tried selecting a save folder as Casper suggested using the same OpenFileDialog but was unable to do so because that Dialog does not have the option to select folders.

KnapSac commented:
@DanielvanDamme we shouldn't keep showing a file dialog in the `continue` case if the user selects no files, instead I think it would be better to treat that as a cancellation, and just return to the home page of step by step. But we can do that next sprint.
